File: iOS/WakeupClock/WakeupClock/Models/AlarmModel.swift
//
//  AlarmModel.swift
//  WakeupClock
//
//  闹钟数据模型
//

import Foundation
import SwiftData

/// 任务类型枚举
enum MissionType: String, Codable, CaseIterable {
    case math = "MATH"
    case shake = "SHAKE"
    case memory = "MEMORY"
    case order = "ORDER"
    case typing = "TYPING"
}

/// 难度等级枚举
enum Difficulty: Int, Codable, CaseIterable {
    case easy = 1
    case medium = 2
    case hard = 3
}

/// 重复模式枚举
enum RepeatMode: String, Codable, CaseIterable {
    case once = "ONCE"
    case workdays = "WORKDAYS"
    case custom = "CUSTOM"
}

/// 闹钟数据模型
@Model
final class AlarmModel {
    /// 唯一标识符
    @Attribute(.unique) var id: String
    
    /// 闹钟时间 (格式: "HH:mm")
    var time: String
    
    /// 是否启用
    var enabled: Bool
    
    /// 标签分类 (work, date, flight, train, meeting, doctor, interview, exam, other)
    var label: String
    
    /// 任务类型
    var missionType: String // 存储为字符串，运行时转换为MissionType
    
    /// 难度等级
    var difficulty: Int // 存储为Int，运行时转换为Difficulty
    
    /// 重复模式
    var repeatMode: String // 存储为字符串，运行时转换为RepeatMode
    
    /// 自定义重复日期 (0=周日, 1=周一, ..., 6=周六)
    var customDays: [Int]
    
    /// 是否跳过节假日
    var skipHolidays: Bool
    
    /// 创建时间
    var createdAt: Date
    
    init(
        id: String = UUID().uuidString,
        time: String,
        enabled: Bool = true,
        label: String = "other",
        missionType: MissionType = .math,
        difficulty: Difficulty = .medium,
        repeatMode: RepeatMode = .workdays,
        customDays: [Int] = [],
        skipHolidays: Bool = false
    ) {
        self.id = id
        self.time = time
        self.enabled = enabled
        self.label = label
        self.missionType = missionType.rawValue
        self.difficulty = difficulty.rawValue
        self.repeatMode = repeatMode.rawValue
        self.customDays = customDays
        self.skipHolidays = skipHolidays
        self.createdAt = Date()
    }
    
    // MARK: - 计算属性（方便访问枚举类型）
    
    var missionTypeEnum: MissionType {
        get { MissionType(rawValue: missionType) ?? .math }
        set { missionType = newValue.rawValue }
    }
    
    var difficultyEnum: Difficulty {
        get { Difficulty(rawValue: difficulty) ?? .medium }
        set { difficulty = newValue.rawValue }
    }
    
    var repeatModeEnum: RepeatMode {
        get { RepeatMode(rawValue: repeatMode) ?? .workdays }
        set { repeatMode = newValue.rawValue }
    }
    
    // MARK: - 辅助方法
    
    /// 获取时间的小时和分钟
    var timeComponents: (hour: Int, minute: Int)? {
        let components = time.split(separator: ":")
        guard components.count == 2,
              let hour = Int(components[0]),
              let minute = Int(components[1]) else {
            return nil
        }
        return (hour, minute)
    }
    
    /// 检查闹钟是否应该在指定日期触发
    func shouldTrigger(on date: Date) -> Bool {
        guard enabled else { return false }
        
        guard let (hour, minute) = timeComponents else { return false }
        
        let calendar = Calendar.current
        let dateComponents = calendar.dateComponents([.year, .month, .day, .hour, .minute, .weekday], from: date)
        
        // 检查时间是否匹配
        guard dateComponents.hour == hour && dateComponents.minute == minute else {
            return false
        }
        
        // 检查重复模式
        switch repeatModeEnum {
        case .once:
            // 只响一次，检查是否是创建日期
            return calendar.isDate(createdAt, inSameDayAs: date)
            
        case .workdays:
            // 工作日：周一到周五
            if let weekday = dateComponents.weekday {
                // weekday: 1=周日, 2=周一, ..., 7=周六
                let isWorkday = (2...6).contains(weekday)
                if !isWorkday { return false }
            }
            
        case .custom:
            // 自定义日期
            if let weekday = dateComponents.weekday {
                // 转换为0=周日, 1=周一, ..., 6=周六的格式
                let dayIndex = weekday == 1 ? 0 : weekday - 1
                if !customDays.contains(dayIndex) { return false }
            }
        }
        
        // 检查是否跳过节假日（使用智能判断，考虑调休上班日）
        if skipHolidays && HolidayChecker.shouldSkipAlarm(date) {
            return false
        }
        
        return true
    }
}

File: Android/WakeupClock/app/src/main/java/com/wakeup/clock/data/model/AlarmModel.kt
package com.wakeup.clock.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.UUID

/**
 * 任务类型枚举
 */
enum class MissionType {
    MATH,
    SHAKE,
    MEMORY,
    ORDER,
    TYPING
}

/**
 * 难度等级枚举
 */
enum class Difficulty(val value: Int) {
    EASY(1),
    MEDIUM(2),
    HARD(3)
}

/**
 * 重复模式枚举
 */
enum class RepeatMode {
    ONCE,
    WORKDAYS,
    CUSTOM
}

/**
 * 闹钟数据模型
 */
@Entity(tableName = "alarms")
data class AlarmModel(
    @PrimaryKey
    val id: String = UUID.randomUUID().toString(),
    
    /** 闹钟时间 (格式: "HH:mm") */
    val time: String,
    
    /** 是否启用 */
    val enabled: Boolean = true,
    
    /** 标签分类 (work, date, flight, train, meeting, doctor, interview, exam, other) */
    val label: String = "other",
    
    /** 任务类型 */
    val missionType: MissionType = MissionType.MATH,
    
    /** 难度等级 */
    val difficulty: Difficulty = Difficulty.MEDIUM,
    
    /** 重复模式 */
    val repeatMode: RepeatMode = RepeatMode.WORKDAYS,
    
    /** 自定义重复日期 (0=周日, 1=周一, ..., 6=周六) */
    val customDays: List<Int> = emptyList(),
    
    /** 是否跳过节假日 */
    val skipHolidays: Boolean = false,
    
    /** 创建时间 */
    val createdAt: Long = System.currentTimeMillis()
) {
    /**
     * 获取时间的小时和分钟
     */
    val timeComponents: Pair<Int, Int>?
        get() {
            val parts = time.split(":")
            if (parts.size != 2) return null
            val hour = parts[0].toIntOrNull() ?: return null
            val minute = parts[1].toIntOrNull() ?: return null
            return Pair(hour, minute)
        }
    
    /**
     * 获取标签对应的图标名称
     */
    val iconName: String
        get() = when (label) {
            "work" -> "briefcase"
            "date" -> "favorite"
            "flight" -> "flight"
            "train" -> "train"
            "meeting" -> "groups"
            "doctor" -> "medical_services"
            "interview" -> "person_add"
            "exam" -> "school"
            else -> "alarm"
        }
    
    /**
     * 获取标签对应的视频名称
     */
    val videoName: String?
        get() = when (label) {
            "work" -> "work"
            "date" -> "date"
            "flight" -> "flight"
            "train" -> "train"
            "meeting" -> "meeting"
            "doctor" -> "doctor"
            "interview" -> "interview"
            "exam" -> "exam"
            else -> null
        }
}
